test/e2e/framework/config.go

```diff
@@ -0,0 +1,299 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"regexp"
+
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/exec"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	// DefaultManagementClusterName is the default name of the Kind cluster
+	// used by the the e2e framework.
+	DefaultManagementClusterName = "mgmt"
+
+	// DefaultKubernetesVersion is the default version of Kubernetes to deploy
+	// for testing.
+	DefaultKubernetesVersion = "v1.16.2"
+)
+
+// ComponentSourceType indicates how a component's source should be obtained.
+type ComponentSourceType string
+
+const (
+	// URLSource is component YAML available directly via a URL.
+	// The URL may begin with file://, http://, or https://.
+	URLSource ComponentSourceType = "url"
+
+	// KustomizeSource is a valid kustomization root that can be used to produce
+	// the component YAML.
+	KustomizeSource ComponentSourceType = "kustomize"
+)
+
+// ComponentSource describes how to obtain a component's YAML.
+type ComponentSource struct {
+	// Name is used for logging when a component has multiple sources.
+	Name string `json:"name,omitempty"`
+
+	// Value is the source of the component's YAML.
+	// May be a URL or a kustomization root (specified by Type).
+	Value string `json:"value"`
+
+	// Type describes how to process the source of the component's YAML.
+	//
+	// Defaults to "kustomize".
+	Type ComponentSourceType `json:"type,omitempty"`
+
+	// Replacements is a list of patterns to replace in the component YAML
+	// prior to application.
+	Replacements []ComponentReplacement `json:"replacements,omitempty"`
+}
+
+// WaiterType indicates the type of check to use to determine if the
+// installed components are ready.
+type WaiterType string
+
+const (
+	// ServiceWaiter indicates to wait until a service's condition is Available.
+	ServiceWaiter WaiterType = "service"
+
+	// PodsWaiter indicates to wait until all the pods in a namespace have a
+	// condition of Ready.
+	PodsWaiter WaiterType = "pods"
+)
+
+// ComponentWaiter contains information to help determine whether installed
+// components are ready.
+type ComponentWaiter struct {
+	// Value varies depending on the specified check Type, ex. when Type is
+	// "service" then Value is the name of a Service resource.
+	Value string `json:"value"`
```

Comment:
I see what is the goal for this. But the values of the « Value » field doesn’t make sense unless we document for each type what is the expected value (for pods and service types)